src/preprocessing.py

```diff
@@ -0,0 +1,443 @@
+
+    # Split the tuple results into separate DataFrame columns
+    issues_df['code_snippets'] = markdown_results.apply(lambda x: x[0])
+    issues_df['images'] = markdown_results.apply(lambda x: x[1])
+    issues_df['links'] = markdown_results.apply(lambda x: x[2])
+    issues_df['tables'] = markdown_results.apply(lambda x: x[3])
+    issues_df['cleaned_body'] = markdown_results.apply(lambda x: x[4])
+
+    print("\nPreprocessing cleaned issue bodies in parallel...")
+    issues_df['classical_preprocessed_body'] = issues_df['cleaned_body'].parallel_apply(preprocess_text_classical)
+
+    return issues_df
+
+def split_data(issues_df: pd.DataFrame, train_range: Tuple[int, int], test_range: Tuple[int, int]) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """
+    Split the issues DataFrame into training and test sets based on specified ranges of GitHub IDs.
+
+    Args:
+        issues_df (pd.DataFrame): The DataFrame containing the issues.
+        train_range (Tuple[int, int]): The range of GitHub IDs for the training set.
+        test_range (Tuple[int, int]): The range of GitHub IDs for the test set.
+
+    Returns:
+        Tuple[pd.DataFrame, pd.DataFrame]:
+            - pd.DataFrame: The training set DataFrame.
+            - pd.DataFrame: The test set DataFrame.
+
+    Raises:
+        ValueError: If 'github_id' column is missing in the DataFrame.
+    """
+    if 'github_id' not in issues_df.columns:
+        raise ValueError("The DataFrame must contain a 'github_id' column.")
+
+    train_start, train_end = train_range
+    test_start, test_end = test_range
+
+    train_set = issues_df[(issues_df['github_id'] >= train_start) & (issues_df['github_id'] <= train_end)]
+    test_set = issues_df[(issues_df['github_id'] >= test_start) & (issues_df['github_id'] <= test_end)]
+    return train_set, test_set
+
+def save_data(dataset_df: pd.DataFrame, path: str) -> None:
+    """
+    Save the DataFrame as a JSON file at the specified path.
+
+    Args:
+        dataset_df (pd.DataFrame): The DataFrame to be saved.
+        path (str): The file path where the JSON file will be saved.
+
+    Returns:
+        None
+
+    Raises:
+        IOError: If the file cannot be written.
+    """
+    try:
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        dataset_df.to_json(path, orient='records', indent=2)
+    except IOError as e:
+        raise IOError(f"Failed to save data to {path}: {e}")
+
+def main() -> None:
+    """
+    Main function to execute the data preprocessing pipeline.
+
+    Steps:
+        1. Pull issues from the specified GitHub repository.
+        2. Remove issues with infrequent assignees.
+        3. Preprocess issues by extracting markdown elements and cleaning text.
+        4. Split data into training and test sets.
+        5. Save the datasets to JSON files.
+
+    Raises:
+        Exception: If any step in the pipeline fails.
+    """
+    try:
+        from pull_issues import pull_issues
```

Comment:
I am importing it here to prevent it from being invoked by the preprocessing tests. Otherwise, I would also need to import it in test_preprocessing.py. I chose this approach to ensure it is only accessible during the actual execution of main. If you have a better suggestion, please feel free to offer your advice.